### PR TITLE
Simplified MS Marco Passage ranking Part + Batch container re-ranking option

### PR DESCRIPTION
**Pull request by jobergum**
- Moved query encodings to use model-evaluation in stateless container which saves resources. Remove query document type which was an error source plus round trips 
- Add ability to re-rank documents in container using batch with model-evaluation 
- Simplify and better instructions 

Ideally should be able to tune intra threads for onnx model evaluations for the 3 different models which are used in the container (https://github.com/vespa-engine/vespa/issues/19084) 

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: msmarco-ranking/src/main/java/ai/vespa/searcher/QueryEncodingSearcher.java [hunk 1]
```diff
@@ -43,15 +50,15 @@ private void federateAndUpdateQuery(Query query, Execution execution, List<Searc
         }
         for (FutureResult f : results) {
             Result r = f.get();
+            if(r.getTotalHitCount() == 0)
+                throw new RuntimeException("Unexpected 0 hits from encoder, this is an server error");
```
Comment (bratseth) [inline msmarco-ranking/src/main/java/ai/vespa/searcher/QueryEncodingSearcher.java:54]:
```suggestion
                throw new RuntimeException("Unexpected 0 hits from encoder, this is a server error");
```

File: msmarco-ranking/src/main/java/ai/vespa/searcher/QueryTensorInput.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import com.yahoo.tensor.TensorType;
+
+/**
+ * Utility to be able to tokenize the query only once and
+ * pass an instance of this class through the searcher chains
+ * See <a href="https://docs.vespa.ai/en/searcher-development.html#passing-information-between-searchers">passing
+ * information between searchers</a>
+ */
+
+@Provides(QueryTensorInput.NAME)
```
Comment (bratseth) [inline msmarco-ranking/src/main/java/ai/vespa/searcher/QueryTensorInput.java:18]:
```suggestion
```

File: msmarco-ranking/passage-ranking.md [hunk 3]
```diff
@@ -128,31 +138,40 @@ schema passage {
...
 
 The *text_token_ids* contains the BERT token vocabulary ids and is only used by the final cross all to all interaction re-ranking model. 
 Storing the token ids in the passage document avoids tokenization at query serving time 
 and we can read the tensor from memory during ranking. 
 
-The *mini_document_embedding* field is the dense vector produced by the sentence encoder model. We enable HNSW indexing for efficient fast 
-approximate nearest neighbor search and we use innerproduct distance metric (query and document vectors normalized to unit length). 
+The *text_token_ids* is an example of a [paged tensor attribute](https://docs.vespa.ai/en/attributes.html#paged-attributes). Using paged attribute 
+enables storing more passages per node, at the cost of potentially slower access time, depending on memory pressure and document access locality. As
+thisf field is only used during re-ranking the number of page-ins are limited by the re-ranking depth.
```
Comment (kkraune) [inline msmarco-ranking/passage-ranking.md:158]:
```suggestion
this field is only used during re-ranking the number of page-ins are limited by the re-ranking depth. 
```

File: msmarco-ranking/passage-ranking.md [hunk 4]
```diff
@@ -128,31 +138,40 @@ schema passage {
...
 and we can read the tensor from memory during ranking. 
 
-The *mini_document_embedding* field is the dense vector produced by the sentence encoder model. We enable HNSW indexing for efficient fast 
-approximate nearest neighbor search and we use innerproduct distance metric (query and document vectors normalized to unit length). 
+The *text_token_ids* is an example of a [paged tensor attribute](https://docs.vespa.ai/en/attributes.html#paged-attributes). Using paged attribute 
+enables storing more passages per node, at the cost of potentially slower access time, depending on memory pressure and document access locality. As
+thisf field is only used during re-ranking the number of page-ins are limited by the re-ranking depth. 
+
+The *mini_document_embedding* field is the dense vector produced by the sentence encoder model. HNSW indexing is enabled for efficient fast 
+approximate nearest neighbor search. As explained earlier the single representation are L2 normalized so one can use the
```
Comment (kkraune) [inline msmarco-ranking/passage-ranking.md:161]: representations?

File: msmarco-ranking/passage-ranking.md [hunk 5]
```diff
@@ -264,48 +287,44 @@ There are several ranking profiles defined in the passage schema and the best pe
...
+The following illustrates a scalable deployment using Vespa where the stateless container cluster 
+handles query parsing and query tokenization. The query encoding models are also evaluated in the stateless container using model evaluation.
+
+ Each content node index a partition of the total document volume and a single query retrieves and rank documents as specified in the
+ ranking profile which happens locally on each node invoved in the query. 
+ In the container one can re-rank the globally top-k documents after obtaining merging the results from the contentn nodes.  
+ Merging might also include diversifying the result set using [Vespa grouping](https://docs.vespa.ai/en/grouping.html)
+ 
+ Moving the most costly ranking model to the stateless layer allows easier auto-scaling to scale with query traffic and as the input size 
+ to the model is limited (e,g 100 hits) the network does not become a bottleneck.
```
Comment (kkraune) [inline msmarco-ranking/passage-ranking.md:320]:
```suggestion
 to the model is limited (e.g. 100 hits) the network does not become a bottleneck. 
```

File: msmarco-ranking/src/main/application/search/query-profiles/dense-colbert-container-rerank.xml [hunk 6]
```diff
@@ -0,0 +1,5 @@
+<!-- Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root. -->
```
Comment (kkraune) [inline msmarco-ranking/src/main/application/search/query-profiles/dense-colbert-container-rerank.xml:1]:
```suggestion
<!-- Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root. -->
```

**Review by bratseth — APPROVED** [review]
Great!

File: msmarco-ranking/src/main/bash/download_models.sh [hunk 7]
```diff
@@ -0,0 +1,19 @@
+#!/bin/bash
+
```
Comment (kkraune) [inline msmarco-ranking/src/main/bash/download_models.sh:2]:
```suggestion
# Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.

```

File: msmarco-ranking/src/main/java/ai/vespa/searcher/QueryTensorInput.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (kkraune) [inline msmarco-ranking/src/main/java/ai/vespa/searcher/QueryTensorInput.java:1]:
```suggestion
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```

File: msmarco-ranking/src/main/java/ai/vespa/searcher/ReRankingSearcher.java [hunk 9]
```diff
@@ -0,0 +1,167 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (kkraune) [inline msmarco-ranking/src/main/java/ai/vespa/searcher/ReRankingSearcher.java:1]:
```suggestion
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```

File: msmarco-ranking/src/test/java/ai/vespa/searcher/QueryTensorInputTest.java [hunk 10]
```diff
@@ -0,0 +1,67 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (kkraune) [inline msmarco-ranking/src/test/java/ai/vespa/searcher/QueryTensorInputTest.java:1]:
```suggestion
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```

File: msmarco-ranking/src/test/java/ai/vespa/searcher/ReRankingSearcherTest.java [hunk 11]
```diff
@@ -0,0 +1,132 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (kkraune) [inline msmarco-ranking/src/test/java/ai/vespa/searcher/ReRankingSearcherTest.java:1]:
```suggestion
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```

File: msmarco-ranking/src/test/java/ai/vespa/searcher/TokenizerFactory.java [hunk 12]
```diff
@@ -0,0 +1,27 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (kkraune) [inline msmarco-ranking/src/test/java/ai/vespa/searcher/TokenizerFactory.java:1]:
```suggestion
// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```

**Review by kkraune — APPROVED** [review]

**Pull request merged**